src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/Completion/TagHelperCompletionProvider.cs

```diff
@@ -71,10 +71,11 @@ public override IReadOnlyList<RazorCompletionItem> GetCompletionItems(RazorCompl
                 containingTagNameToken.Span.IntersectsWith(context.AbsoluteIndex))
             {
                 if ((containingTagNameToken.FullWidth > 1 || containingTagNameToken.Content == "-") &&
-                    containingTagNameToken.Span.Start != context.AbsoluteIndex)
+                    containingTagNameToken.Span.Start != context.AbsoluteIndex &&
+                    containingTagNameToken.Span.End != context.AbsoluteIndex)
```

Comment:
Hmmm, I believe this will result in partially written attributes getting "only" TagHelper completion. On paper this may seem ideal; however, typing out something like `onclick` may result in only the TagHelper variants of completions because of this check. I like your intent, do things still operate correctly if you instead modify these checks to take into account the completion reason? Aka if it's not typing, allow completion: https://github.com/dotnet/razor-tooling/blob/0ba299a277c9d4a8be693ce3dda7054232b976e9/src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/Completion/RazorCompletionContext.cs#L14

Reply:
> typing out something like `onclick` may result in only the TagHelper variants of completions because of this check

I don't understand how that can happen. This completion provider just returns its items, then they are combined with items from html server, how can some internal logic of one provider change the output of a whole LSP server? Also I am a bit confused, that you propose `onclick` while pointing to the lines, that are for tag helpers, not attributes. I created a simple test tag helper:
```cs
[HtmlTargetElement("input", Attributes = "onclick")]
public class TestInputTagHelper : TagHelper
{
    [HtmlAttributeName("onclick")]
    public int OnClick { get; set; }
}
```
And html completions are not gone:
![r2XrrgvqhN](https://user-images.githubusercontent.com/70431552/178579663-8587b501-edeb-498d-8bc1-a59ede665914.gif)
I guess this was scenario you worried about

> do things still operate correctly if you instead modify these checks to take into account the completion reason?

Yes, I verified, that if a change the given condition to `context.Reason != CompletionReason.Invoked` this will work. However, I don't see a logical connection between these two facts. It's just happened to be, that in our case it will lead us to the same result. IMO check for the last position in the html tag/minimalized attribute is a lot more clear, because it is far better readable: we are checking if we are in the last position of (in this case) html tag, and if we are, we want to bypass this block and still get completions from the code below. Adding completion reason as an additional check doesn't seem to be required here as well, because it will be kind of redundant, as it doesn't contribute to the final result.

If I somehow misunderstood you, please provide an example, that will require to bring the check for completion reason. I do not see such yet.

Reply:
AHh so in your case the reason why your gif is working as expected is that the completion session started when you typed `o` and then the items are just being filtered down. Try doing ctrl + space at `onc|` and I'm betting with your change you'll see only the TagHelper variant come through

Reply:
Tag helper:
```cs
[HtmlTargetElement("input", Attributes = "onclick")]
public class TestInputTagHelper : TagHelper
{
    [HtmlAttributeName("onclick")]
    public int OnClick { get; set; }
}
```

Completions:
![GBst95myDj](https://user-images.githubusercontent.com/70431552/178664874-d6c015bd-875a-4fe4-ace1-c7806ff5e072.gif)
Problem I see here is that we don't offer tag helper attribute completion, while we should, because it is a "potential tag helper" case. I'll fix that

Now remove requirement to have an `onclick` attribute for `input` in order to be a tag helper:
```cs
[HtmlTargetElement("input")]
public class TestInputTagHelper : TagHelper
{
    [HtmlAttributeName("onclick")]
    public int OnClick { get; set; }
}
```

Now completions are these:
![VTxTBJHaAf](https://user-images.githubusercontent.com/70431552/178665768-bc17bc31-990d-4f5d-ab9b-aeba9aa7cc86.gif)
The behaviour is expected

Reply:
No 😛 (again, both variants):
![u92wON8tZg](https://user-images.githubusercontent.com/70431552/178801051-496c990c-2d2e-429d-8458-81d5e53fcf42.gif)

![FOoFDkAy7F](https://user-images.githubusercontent.com/70431552/178801111-f231f452-1525-4c7e-9fb3-b0e69bce8a73.gif)

Reply:
lmao, I am thoroughly confused as to how that's happening. You know what, lets take this and roll with it. If it causes any sort of problem we can always revisit. Love that you stayed with me through all this back and forth. I'm sure I was a bit of a bother ❤️